public/app/plugins/datasource/prometheus/result_transformer.ts

```diff
@@ -1,234 +1,262 @@
-      throw new Error('Prometheus heatmap error: data should be a time series');
-    }
+  const dataFrame: DataFrame[] = [];
+
+  prometheusResult.result.forEach((data: MatrixOrVectorResult) => dataFrame.push(transformToDataFrame(data, options)));
+
+  if (options.format === 'heatmap') {
+    dataFrame.sort(sortSeriesByLabel);
+    const seriesList = transformToHistogramOverTime(dataFrame);
+    return seriesList;
+  }
+
+  return dataFrame;
+}
+
+/**
+ * Transforms matrix and vector result from Prometheus to DataFrame
+ */
+function transformToDataFrame(data: MatrixOrVectorResult, options: TransformOptions): DataFrame {
+  const { name } = createLabelInfo(data.metric, options);
 
-    for (const value of metricData.values) {
+  const fields: Field[] = [];
+
+  if ('values' in data) {
+    const stepMs = options.step ? options.step * 1000 : NaN;
+    let baseTimestamp = options.start * 1000;
+    const dps: PromValue[] = [];
+
+    for (const value of data.values) {
       let dpValue: number | null = parseFloat(value[1]);
 
-      if (_.isNaN(dpValue)) {
+      if (isNaN(dpValue)) {
         dpValue = null;
       }
 
-      const timestamp = parseFloat(value[0]) * 1000;
+      const timestamp = value[0] * 1000;
       for (let t = baseTimestamp; t < timestamp; t += stepMs) {
-        dps.push([null, t]);
+        dps.push([t, null]);
       }
       baseTimestamp = timestamp + stepMs;
-      dps.push([dpValue, timestamp]);
+      dps.push([timestamp, dpValue]);
     }
 
-    const endTimestamp = end * 1000;
+    const endTimestamp = options.end * 1000;
     for (let t = baseTimestamp; t <= endTimestamp; t += stepMs) {
-      dps.push([null, t]);
+      dps.push([t, null]);
     }
+    fields.push(getTimeField(dps, true));
+    fields.push(getValueField(dps, undefined, false));
+  } else {
+    fields.push(getTimeField([data.value]));
+    fields.push(getValueField([data.value]));
+  }
 
+  return {
+    meta: options.meta,
+    refId: options.refId,
+    length: fields[0].values.length,
+    fields,
+    name,
+  };
+}
+
+function transformMetricDataToTable(md: MatrixOrVectorResult[], options: TransformOptions): DataFrame {
+  if (!md || md.length === 0) {
     return {
-      datapoints: dps,
-      refId: options.refId,
-      target: name ?? '',
-      tags: labels,
-      title,
       meta: options.meta,
+      refId: options.refId,
+      length: 0,
+      fields: [],
     };
   }
 
-  transformMetricDataToTable(
-    md: any,
-    resultCount: number,
-    refId: string,
-    meta: QueryResultMeta,
-    valueWithRefId?: boolean
-  ): TableModel {
-    const table = new TableModel();
-    table.refId = refId;
-    table.meta = meta;
-
-    let i: number, j: number;
-    const metricLabels: { [key: string]: number } = {};
-
-    if (!md || md.length === 0) {
-      return table;
+  const metricLabels: string[] = [];
+  // Collect all labels across all metrics
+  md.forEach(series => {
+    for (const label in series.metric) {
```

Comment:
this should probably be `const label of series.metric`

Reply:
series.metric is an object here so for of wouldn't work.

Reply:
hmm in that case maybe it should also have `hasOwnProperty` check? Or just use `Object.keys`? I assume it's not an issue at this moment but not sure if we should count on how the object is created.

Also in general I think you could just do something like:
```
Object.keys(md.reduce((acc, series) => ({...acc, ...series}) ,{})).sort().map(...)
// or
_.uniq(md.flatMap((series) => Object.keys(series))).sort().map()
```

To get the unique `metricLabels`

Reply:
nice, I like the Object.keys way more.